### ISSUE TITLE
Rewrite the downloader

### ISSUE THREAD

**Issue by arkon** [tachiyomiorg/tachiyomi#7492]
There's a bunch of issues with existing implementation, partially due to weird RxJava behaviour.

- #959: auto restart downloads
- #1501: Remove .tmp folder after download is cancelled/failed
- #5010: download queue still download deleted chapter
- #5503: Download pauses when the last chapter in queue is removed
- #5477: Download Queue Unpauses Itself Whenever Something is Added to the Queue While Paused
- #6002: "Download completed" notification shown after cancelling or paused due to network issues
- #6442: Download remaining in queue if you finish a chapter just as the cbz file is being created
- #6532: Completion bar does not correspond with pages downloaded
- #6871: Downloads dont start on their own
- #7249: Download queue reordering causes redownload
- #7258: Downloading pages in parallel does not work?

**Comment by AntsyLich** [comment]
Seems related #6442, #5966, #1561

**Comment by iMGDE** [comment]
Feature Request (since related to this didnt open new issue): please add multithreaded downloads in the sense of multiple chapters/sources at once

**Comment by AntsyLich** [comment]
> multiple chapters/sources at once 

Consecutive chapter download from same source won't be added. Multiple source is capped at 5 at a time. So you will have maximum of 5 downloads running considering you have 5 or more sources in download queue.

> (cause at least for me it only downloads from 2 sources)

Current downloader has a bug which makes it that if a source has more than 20 chapters in queue chapters from the next source doesn't start downloading.

**Comment by scb261** [comment]
Not sure if it can be handled by the app, but sometimes user's system appends `(1)` to folder names and it becomes unreadable for the app. Reasons:
- not allowed characters https://github.com/tachiyomiorg/tachiyomi-extensions/issues/5191
- different capitalization https://github.com/tachiyomiorg/tachiyomi-extensions/issues/3536, https://github.com/tachiyomiorg/tachiyomi-extensions/issues/6284
- not even clear why it happened https://github.com/tachiyomiorg/tachiyomi-extensions/issues/10554#issuecomment-1240813383
